build-aux/snap/snapcraft.yaml

```diff
@@ -28,172 +29,272 @@ license: GPL-3.0
+      install -Dm644 -t "${CRAFT_PART_INSTALL}/usr/lib/snapd/apparmor" parser.conf
+      cd "${CRAFT_PART_BUILD}/profiles"
+      make -j"${CRAFT_PARALLEL_BUILD_COUNT}"
+      mkdir -p "${CRAFT_PART_INSTALL}/usr/lib/snapd/apparmor.d"
+      cp -a apparmor.d/abi "${CRAFT_PART_INSTALL}/usr/lib/snapd/apparmor.d"
+      cp -a apparmor.d/abstractions "${CRAFT_PART_INSTALL}/usr/lib/snapd/apparmor.d"
+      cp -a apparmor.d/tunables "${CRAFT_PART_INSTALL}/usr/lib/snapd/apparmor.d"
+    prime:
+      # We are statically linked
+      - -usr/lib/$CRAFT_ARCH_TRIPLET_BUILD_FOR/**
+      - -usr/include
+
+  snapd:
+    plugin: nil
+    source: .
+    build-snaps:
+      - go/1.18/stable
+    after:
+      - apparmor
+    # these packages are needed to call mkversion.sh in override-pull, all other
+    # dependencies are installed using apt-get build-dep
+    build-packages:
+      - git
+      - dpkg-dev
+      - autoconf
+      - autoconf-archive
+      - automake
+      - xfslibs-dev
+      - libudev-dev
+      - libcap-dev
+      - libseccomp-dev
+      - pkg-config
+    override-pull: |
+      craftctl default
+      # set version, this needs dpkg-parsechangelog (from dpkg-dev) and git
+      craftctl set version="$(./mkversion.sh --output-only)"
+      ./get-deps.sh --skip-unused-check
+    override-build: |
+      ./mkversion.sh
+      cd "${CRAFT_PART_BUILD}/cmd"
+
+      autoreconf -fvi
+      ./configure --prefix=/usr --libexec=/usr/lib/snapd --without-unit-tests --enable-static-libapparmor
+      make -j"${CRAFT_PARALLEL_BUILD_COUNT}"
+      make -j1 DESTDIR="${CRAFT_PART_INSTALL}" install
+
+      cd "${CRAFT_PART_BUILD}"
+
+      # Work-around for snapcraft not working with git worktrees
+      # We detect it with .git being a file rather than a directory
+      if [ -f .git ]; then
+        EXTRA_GO_FLAGS="-buildvcs=false"
       fi
-      make -j$(nproc)
-      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/lib/snapd
-      cp -a apparmor_parser $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/
-      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor
-      cp -a parser.conf $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor/
-      cd $SNAPCRAFT_PART_BUILD/profiles
-      make -j$(nproc)
-      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor.d
-      cp -a apparmor.d/abi $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor.d/
-      cp -a apparmor.d/abstractions $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor.d/
-      cp -a apparmor.d/tunables $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor.d/
+      CMDS=(bin/snap
+            lib/snapd/snapd
+            lib/snapd/snap-exec
+            lib/snapd/snap-failure
+            lib/snapd/snap-fde-keymgr
+            lib/snapd/snap-preseed
+            lib/snapd/snap-recovery-chooser
+            lib/snapd/snap-repair
+            lib/snapd/snap-seccomp
+            lib/snapd/snap-update-ns
+            lib/snapd/snapctl
+            lib/snapd/snapd-apparmor)
+      for cmd in "${CMDS[@]}"; do
```

Comment:
We should really resurrect the effort to move the logic we have to the makefile, so that we don't have to have the crazy-complex build rules duplicated per packaging system and really mostly not kept in sync except when things explode.

We should have build profiles and only allow the selection of a profile, everything else should be baked in.

Reply:
Agreed.

Reply:
I'd love to pick this up soon.